File: rust/tests/device_type_entity_test.rs
// Generated basic-flow test for the device_type entity (model-driven;
// mirrors the go TestEntity generator).

#![allow(unused_variables, unused_mut, unused_imports)]

mod common;

use std::rc::Rc;

use common::*;

use bluefin_decryptx_p2pe_sdk::core::helpers::{getp, getpath, ja, jo, now_ms, setp, to_map};
use bluefin_decryptx_p2pe_sdk::utility::voxgigstruct as vs;
use bluefin_decryptx_p2pe_sdk::{test_sdk, Entity, BluefinDecryptxP2peEntity, BluefinDecryptxP2peSDK, Value};

#[test]
fn device_type_entity_instance() {
    let testsdk = test_sdk(Value::Noval, Value::Noval);
    let ent = testsdk.device_type(Value::Noval);
    assert_eq!(ent.get_name(), "device_type");
}

#[test]
fn device_type_entity_stream() {
    // stream() runs the list op through the full pipeline and yields each
    // result item. Seed two entities via test mode; with the `streaming`
    // feature active it yields the feature's incremental items, else it
    // falls back to the materialised items — either way every item yields.
    let seed = jo(vec![(
        "entity",
        jo(vec![(
            "device_type",
            jo(vec![
                ("strm01", jo(vec![("id", Value::str("strm01"))])),
                ("strm02", jo(vec![("id", Value::str("strm02"))])),
            ]),
        )]),
    )]);

    let sdkopts = jo(vec![(
        "feature",
        jo(vec![("streaming", jo(vec![("active", Value::Bool(true))]))]),
    )]);

    let testsdk = test_sdk(seed.clone(), sdkopts);
    let ent = testsdk.device_type(Value::Noval);
    let items: Vec<Value> = ent
        .stream("list", Value::empty_map(), Value::empty_map())
        .expect("stream failed")
        .collect();
    assert_eq!(items.len(), 2, "stream should yield both seeded items");

    // Fallback: streaming inactive still yields both materialised items.
    let plainsdk = test_sdk(seed, Value::Noval);
    let plainent = plainsdk.device_type(Value::Noval);
    let plain_items: Vec<Value> = plainent
        .stream("list", Value::empty_map(), Value::empty_map())
        .expect("stream failed")
        .collect();
    assert_eq!(plain_items.len(), 2, "fallback stream should yield both items");
}

#[test]
fn device_type_entity_basic() {
    let setup = device_type_basic_setup(Value::Noval);
    // Per-op sdk-test-control.json skip — the basic test exercises a flow
    // with multiple ops; skipping any op skips the whole flow.
    let mode = if setup.live { "live" } else { "unit" };
    for op in ["list", "load"] {
        let (skip, reason) = is_control_skipped("entityOp", &format!("device_type.{}", op), mode);
        if skip {
            let reason = if reason.is_empty() {
                "skipped via sdk-test-control.json".to_string()
            } else {
                reason
            };
            eprintln!("skip: {}", reason);
            return;
        }
    }
    // The basic flow consumes synthetic IDs from the fixture. In live mode
    // without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only {
        eprintln!("skip: live entity test uses synthetic IDs from fixture — set BLUEFIN_DECRYPTX_P2PE_TEST_DEVICE_TYPE_ENTID JSON to run live");
        return;
    }
    let client = setup.client.clone();

    // Bootstrap entity data from existing test data (no create step in flow).
    let device_type_ref01_data_raw = vs::items(&to_map(&getpath(&["existing", "device_type"], &setup.data)));
    let device_type_ref01_data = to_map(&vs::get_elem(
        &vs::get_elem(&device_type_ref01_data_raw, &Value::Num(0.0), Value::Noval),
        &Value::Num(1.0),
        Value::Noval,
    ));
    // LIST
    let device_type_ref01_ent = client.device_type(Value::Noval);
    let device_type_ref01_match = Value::empty_map();

    let device_type_ref01_list = device_type_ref01_ent
        .list(device_type_ref01_match.clone(), Value::Noval)
        .expect("list failed");
    // list resolves to one ENTITY per record; the flow asserts on the
    // records, so map each through data().
    let device_type_ref01_list = ja(device_type_ref01_list.iter().map(|e| e.data(None)).collect::<Vec<Value>>());

    // LOAD
    let device_type_ref01_match_dt0 = jo(vec![("id", getp(&device_type_ref01_data, "id"))]);
    let device_type_ref01_data_dt0_loaded = device_type_ref01_ent
        .load(device_type_ref01_match_dt0.clone(), Value::Noval)
        .expect("load failed");
    let device_type_ref01_data_dt0_load_result = to_map(&device_type_ref01_data_dt0_loaded.data(None));
    assert!(
        matches!(device_type_ref01_data_dt0_load_result, Value::Map(_)),
        "expected load result to be a map"
    );
    assert_eq!(
        getp(&device_type_ref01_data_dt0_load_result, "id"),
        getp(&device_type_ref01_data, "id"),
        "expected load result id to match"
    );

}

fn device_type_basic_setup(extra: Value) -> EntityTestSetup {
    load_env_local();

    let mut entity_data_file = manifest_dir();
    entity_data_file.push("..");
    entity_data_file.push(".sdk");
    entity_data_file.push("test");
    entity_data_file.push("entity");
    entity_data_file.push("device_type");
    entity_data_file.push("DeviceTypeTestData.json");

    let entity_data = read_json(&entity_data_file);

    let options = jo(vec![("entity", getp(&entity_data, "existing"))]);

    let client = test_sdk(options, extra.clone());

    // Generate idmap via transform, matching the TS pattern.
    let idmap = vs::transform(
        &ja(vec![Value::str("device_type01"), Value::str("device_type02"), Value::str("device_type03")]),
        &jo(vec![(
            "`$PACK`",
            ja(vec![
                Value::str(""),
                jo(vec![
                    ("`$KEY`", Value::str("`$COPY`")),
                    (
                        "`$VAL`",
                        ja(vec![
                            Value::str("`$FORMAT`"),
                            Value::str("upper"),
                            Value::str("`$COPY`"),
                        ]),
                    ),
                ]),
            ]),
        )]),
        None,
    )
    .unwrap_or_else(|_| Value::empty_map());

    // Detect ENTID env override before env_override consumes it. When live
    // mode is on without a real override, the basic test runs against
    // synthetic IDs from the fixture and 4xx's.
    let entid_env_raw = std::env::var("BLUEFIN_DECRYPTX_P2PE_TEST_DEVICE_TYPE_ENTID").unwrap_or_default();
    let idmap_overridden =
        !entid_env_raw.trim().is_empty() && entid_env_raw.trim().starts_with('{');

    let env = env_override(jo(vec![
        ("BLUEFIN_DECRYPTX_P2PE_TEST_DEVICE_TYPE_ENTID", idmap.clone()),
        ("BLUEFIN_DECRYPTX_P2PE_TEST_LIVE", Value::str("FALSE")),
        ("BLUEFIN_DECRYPTX_P2PE_TEST_EXPLAIN", Value::str("FALSE")),
        ("BLUEFIN_DECRYPTX_P2PE_APIKEY", Value::str("NONE")),
    ]));

    let idmap_resolved = match to_map(&getp(&env, "BLUEFIN_DECRYPTX_P2PE_TEST_DEVICE_TYPE_ENTID")) {
        Value::Map(m) => Value::Map(m),
        _ => to_map(&idmap),
    };

    let live = getp(&env, "BLUEFIN_DECRYPTX_P2PE_TEST_LIVE") == Value::str("TRUE");

    let client = if live {
        let merged = vs::merge(
            &ja(vec![jo(vec![("apikey", getp(&env, "BLUEFIN_DECRYPTX_P2PE_APIKEY"))]), extra]),
            None,
        );
        BluefinDecryptxP2peSDK::new(to_map(&merged))
    } else {
        client
    };

    EntityTestSetup {
        client,
        data: entity_data,
        idmap: idmap_resolved,
        env: env.clone(),
        explain: getp(&env, "BLUEFIN_DECRYPTX_P2PE_TEST_EXPLAIN") == Value::str("TRUE"),
        live,
        synthetic_only: live && !idmap_overridden,
        now: now_ms(),
    }
}

File: scala/sdktest/DeviceTypeEntityTest.scala
// Generated basic-flow test for the device_type entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped DeviceTypeTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefindecryptxp2pesdk.core.{Helpers, SdkEntity, BluefinDecryptxP2peSDK}
import voxgig.bluefindecryptxp2pesdk.utility.struct.Struct

object DeviceTypeEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("device_type.instance") {
      val testsdk = BluefinDecryptxP2peSDK.testSDK()
      val ent = testsdk.deviceType(null)
      rep.check("device_type.instance", ent != null, "expected non-null device_type entity")
    }

    rep.scope("device_type.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/device_type/DeviceTypeTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinDecryptxP2peSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("device_type01", "DEVICE_TYPE01")
      idmap.put("device_type02", "DEVICE_TYPE02")
      idmap.put("device_type03", "DEVICE_TYPE03")
      val now = System.currentTimeMillis()
      val deviceTypeRef01DataRaw = Struct.items(Helpers.toMapAny(
          Struct.getpath(entityData, "existing.device_type")))
      val deviceTypeRef01Data = Helpers.toMapAny(deviceTypeRef01DataRaw.get(0).get(1))

      // LIST
      val deviceTypeRef01Ent = client.deviceType(null)
      val deviceTypeRef01Match = new LinkedHashMap[String, Object]()
      val deviceTypeRef01ListResult = deviceTypeRef01Ent.list(deviceTypeRef01Match, null)
      rep.check("device_type.list.islist", deviceTypeRef01ListResult.isInstanceOf[JList[?]], "expected list result to be an array, got " + deviceTypeRef01ListResult)

      // LOAD
      val deviceTypeRef01MatchDt0 = new LinkedHashMap[String, Object]()
      deviceTypeRef01MatchDt0.put("id", deviceTypeRef01Data.get("id"))
      val deviceTypeRef01DataDt0Loaded = deviceTypeRef01Ent.load(deviceTypeRef01MatchDt0, null)
      val deviceTypeRef01DataDt0LoadResult = Helpers.toMapAny(deviceTypeRef01DataDt0Loaded match { case e: SdkEntity => e.data(); case o => o })
      rep.check("device_type.load.map", deviceTypeRef01DataDt0LoadResult != null, "expected load result to be a map")
      rep.eq("device_type.load.id", deviceTypeRef01Data.get("id"), deviceTypeRef01DataDt0LoadResult.get("id"))
    }
  }
}
